# Profitable App Profiles for the App Store and Google Play Markets

## 1. Analyzing Mobile App Data

### Skills Learned :
. Basics Python Programming
. List and for-loops
. Conditional statements(if, else, and ifelse)
. Dictionaries and Frequency Tables
. Functions
. Jupyter Notebook functionality

### Objective:
Learn to combine skills to perfomr practical data analysis.

### Project Background:
                                  

### Goal:
Analyze Android and iOS data from Google Play and the App store to help our developers understand what type of apps are likely to attract more users.

## 2. Opening and Exploring the data

In order to analyze what apps will attract more users to google play and app store we need to collect and anlyze data about mobile apps available on those platforms. There are approximately 2million ios apps and 2.1million google play apps. That's too much data too analyze at the moment. So we'll focus on a sample of the data instead.

We'll first see if there exist any relevant existing data that's freely available. There is are two data sets that would be suitable. A data set formm the google play that has ~10,000 android apps from google play coolected in August 2018. Another data set that has ~7,000 ios apps from the app store collected in july 2017.

We'll open and explore the two data sets.


In [1]:
# We start by creating a function that can repeatedly used to print
# rows in a readable fashion.

def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start: end]
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        

In [2]:
# We now want to open the two data sets mentioned above, and save both as lists of lists.

from csv import reader

#Google Play dataset
opened_file = open('/home/lumenetix/Documents/python/data_quest/project1/google_play_data/googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

#iOS Store dataset
opened_file = open('/home/lumenetix/Documents/python/data_quest/project1/apple_app_store_data/AppleStore.csv')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

In [4]:
# We will now print column names and try to identfiy column names
# that will be useful for analysis as well as explore a few lines of each dataset.

print(android_header)
print('\n')
explore_data(android, 0, 6, True)

print(ios_header)
print('\n')
explore_data(ios, 0, 6, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Eve

## 3. Deleting Wrong Data

We need to make sure the data we analyze is accurate. We do this by:

.Detect inaccurate data, and correct or remove it.

.Detect duplicate data, and remove the duplicates.

Since we're concerned with free English speaking apps we can remove non-free apps and non-English apps

This process of preparing our data for analysis is called data cleaning. Data cleaning is done before the analysis; it includes removing or correcting wrong data, removing duplicate data, and modifying the data to fit the purpose of our analysis.



In [7]:
# There is an error for index 10472
print(android_header)
print('/n')
print(android[10472])


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
/n
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [9]:
# We see that the value for Content Rating is '' and the value for 
# Genres is missing

# We're going to delete the row to make the analysis easier
del android[10472]

# We reprint the same index to check if we've properly removed the row
print(android[10472])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


## 4. Removing Duplicate Entries: Part One

There exist duplicate entries in the list so we need to remove them in order to not double count.


In [11]:
# We're going to get any app names that have duplicates

#We create two lists: one for storing the names of duplicates
# and one for storing the names of unique apps.
duplicate_apps = []
unique_apps = []

# we loop through android data set and for each iteration:
# 1. we save the app name to a vriable named name
# 2. if name already exist in unique apps then we append name to
# duplicate_apps list
# 3. else  we append name to the unique_apps list
for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:15])

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


We need to remove the duplicates but we won't do it randomly. Instead we'll remove the the duplicates with the least number of reviews. This we have the largest a better average to work with.

In [14]:
# We'll print a few duplicates just to see the difference

for app in android:
    name = app[0]
    if name == 'Quick PDF Scanner + OCR FREE':
        print(app)

for app in android:
    name = app[0]
    if name == 'Box':
        print(app)


['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80804', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Fr

## 5. Removing Duplicates Entries: Part Two

To remove the duplicates we need to:
1. create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app.
2. use the info stored in the dictionary and creat a new data set, which will have only one entry per app.

In [22]:
# we will use the in and not in operators to check for membership of duplicates

# The expected length
print(len(android)-len(duplicate_apps))
# empty dictionary
reviews_max = {}
#loop through data set
for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

        

# Now we'll move on to removing the duplicate rows:

# We create two empty lists:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews= float(app[3])
    
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)
        
print(android_header)
print('/n')
print(android_clean[:5])
print(len(android_clean))

9659
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
/n
[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up'], ['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Ev

## 6. Removing Non-English Apps: Part One

Since we're concerned with only English apps so we need to remove the non-english ones.

We know that the numbers corresponding to the characters we commonly use in English text are all in the range of 0 to 127 according to ASCII system. So if the number of a character in the name of the app has a value greater than 127 then we can assume it is a non-enlglish character.



In [25]:
# We want to write a function that takes in a string and returns False
# if there's any chracter in the string that doesn't belong to the set
# of common English characters, otherwise it returns True.


def english_check_proto(string):    
    
    #iterate over a string and check each value of each character in it
    for char in string:
        if ord(char) > 127:
            return False
    
    return True


# We're going to check to see if our function works

english_check_proto('Instagram')
english_check_proto('爱奇艺PPS -《欢乐颂2》电视剧热播')
english_check_proto('Docs To Go™ Free Office Suite')
english_check_proto('Instachat 😜')

# From our results we see that we need to account for emojis and
# characters like ™ 



False

## 7. Removing Non-English Apps: Part Two

As noticed before we need to account for certain special characters like ™ and emojis. We can modify our function to check for at least three non-English characters in case we stumble upon the special characters that are ok to have in a name.



In [27]:
# We write the modified function take accept 3 non english characters
# before it returns False
def english_check(string):
    
    counter = 0
    for char in string:
        if ord(char) > 127:
            counter += 1
        if counter > 3:
            return False
        
    return True

english_check('Instagram')
english_check('爱奇艺PPS -《欢乐颂2》电视剧热播')
english_check('Docs To Go™ Free Office Suite')
english_check('Instachat 😜')

# This seems to be a reasonable way to define the function

True

In [32]:
# We're going to use our new function to filter out the non-English
# apps from both data sets.
android_english = []

# We need to loop through a data set.
for app in android_clean:
    name = app[0] 
    if english_check(name): #check if name is English
        android_english.append(app) # Append to new list if it is
        
print(android_english[:5])
print(len(android_english))


[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up'], ['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']]
9614


In [33]:
# We're not going to do the same with iOS data set

ios_english = []

for app in ios:
    name = app[2]
    if english_check(name):
        ios_english.append(app)
        
print(ios_english[:5])
print(len(ios_english))

[['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1'], ['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1'], ['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1'], ['4', '282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262241', '649', '4', '4.5', '5.10.0', '12+', 'Shopping', '37', '5', '9', '1'], ['5', '282935706', 'Bible', '92774400', 'USD', '0', '985920', '5320', '4.5', '5', '7.5.1', '4+', 'Reference', '37', '5', '45', '1']]
6183


## 8. Isolating thre Free Apps

So far we've removed:
1. inaccacurate data
2. duplicate app entries
3. non-English apps

We now need to remove non-free apps.

In [34]:
# We want to loop through each data set to isolate the fee apps in
# separate lists.

print(android_header)
print(ios_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [42]:
# 7 and 5 are the respective indices for the price column of 
# the android and ios data sets

android_free = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_free.append(app)
        
print(android_free[:5])
print(len(android_free))

[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up'], ['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']]
8864


In [41]:
# We do the same with the ios data set

ios_free = []

for app in ios_english:
    price = app[5]
    if price =='0':
        ios_free.append(app)
        
print(ios_free[:5])

print(len(ios_free))

[['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1'], ['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1'], ['4', '282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262241', '649', '4', '4.5', '5.10.0', '12+', 'Shopping', '37', '5', '9', '1'], ['5', '282935706', 'Bible', '92774400', 'USD', '0', '985920', '5320', '4.5', '5', '7.5.1', '4+', 'Reference', '37', '5', '45', '1'], ['7', '283646709', 'PayPal - Send and request money safely', '227795968', 'USD', '0', '119487', '879', '4', '4.5', '6.12.0', '4+', 'Finance', '37', '0', '19', '1']]
3222


## 9. Most Common Apps by Genre: Part One

Now that we've cleaned our data sets we can move towards trying to determine the kinds of apps that are likely to attract more users.

We want to minimize risk and overhead, so our validation strategy will be comprosed of 3 steps:
1. build aminimal android verions of the app, and add it to google play.
2. if app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build and ios version of the app and add it to the app store.

An app profile that has been succesful in both google play and app store is ideal. We can start by see which genres of apps are most common. To do this we need to use frequency tables for a few columns in our data sets

In [48]:
print(android_header)
print('\n')
print(android_free[:5])
print('\n')
print(ios_header)
print('\n')
print(ios_free[:5])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up'], ['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone

We can see that the columns that are useful for android data set are the columns 1 and 9. For the ios data set the column index we care about is 12


## 10. Most Common Apps by Genre: Part Two

So from looking through the data sets we see that the Genres and Category columns are useful for the google play data set and the prime_genre column is useful from the App Store data set.

We can build two functions to help analyze the frequency tables:
1. a function to generate frequency tables that show percentages
2. a function to display the percentages in a descending order



In [52]:
# we first start by creating a helper function which transforms a
# frequency table into a list of tuples, then sorts the list in a
# descending order

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
    

In [53]:
# Now we create a function that takes a dataset and index
# and produces a frequency table any column.

def freq_table(dataset, index):
    table = {}
    app_count = 0
    
    for app in dataset:
        key = app[index]
        app_count += 1
        if key in table:
            table[key] += 1
        else:
            table[key] = 1
    
    percentage_table = {}
    for key in table:
        percentage = (table[key] / app_count) * 100
        percentage_table[key] = percentage
        
    return percentage_table

In [58]:
# frequency table for category in android
display_table(android_free, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [59]:
# frequency table for genres in android
display_table(android_free, 9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

In [60]:
# frequency table for prime_genre in ios
display_table(ios_free, 12)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


## 11. Most Common Apps by Genre: Part Three

Now that we have frequency tables generated we can focus on analyzing them.

We must note first that we are only including free English apps. So our analysis should not extend beyond that.

We start by analyzing the frequency table generated for prime_genre in
the App Store data set. The most common genre is Games at 58.16% and the runner up is Entertainment at %7.88.

Most of the apps are not necessarily designed for practicality. Most of the apps seems to be geared towards leasure activities. A quick assesment of this data by itself would lead me to think that a entertainment like profile would be effective in getting more users.

Although there are a large number of apps for games this doesn't necessarily imply that games also have the most users. But then again if games were as effective at attracting users then it wouldn't be a popular type of app to produce because it wouldn't necessarily be profitable. From what I know about games is that they are designed to be as addicting as possible so that users use them as much as possible and there is incentive to invite other users.



We now look at the frequency tables for category and genres columns of the google play data set.

Looking at the genres data we have the top three:
Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725

Looking at the category data we have:
FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072

The pattern is a bit different compared to ios data. This doesn't make it as obvious as to what king of app profile would be attractive to as many users as possible.

The most obvious thing from these results is that entertainment type apps appear to be more abundent.

## 12. Most Popular Apps by Genre on the App Store

To reiterate the data seems to imply that the stores are dominated by apps that are made for fun.

We now need to find out which apps have the most users. One way to do this is to calculate the average number of installs for each app genre. For Google Play we can just use the installs columns but for the App Store we don't have such column so we instead could use rating_count_tot.

We need to start by calculating the average number of users per app genre. That means we need to:
1. isolate the apps of each genre
2. sum up the user ratings for the apps of that genre
3. divide the sum by the number of apps belonging to that genre

In [62]:
# We need to generate a frequency table for prime_genre column to
# get the unique app genres.

ft_prime_genre = freq_table(ios_free, -5)



for genre in ft_prime_genre:
    total = 0
    len_genre = 0
    for app in ios_free:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[6])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)


Productivity : 21028.410714285714
Weather : 52279.892857142855
Shopping : 26919.690476190477
Reference : 74942.11111111111
Finance : 31467.944444444445
Music : 57326.530303030304
Utilities : 18684.456790123455
Travel : 28243.8
Social Networking : 71548.34905660378
Sports : 23008.898550724636
Health & Fitness : 23298.015384615384
Games : 22788.6696905016
Food & Drink : 33333.92307692308
News : 21248.023255813954
Book : 39758.5
Photo & Video : 28441.54375
Entertainment : 14029.830708661417
Business : 7491.117647058823
Lifestyle : 16485.764705882353
Education : 7003.983050847458
Navigation : 86090.33333333333
Medical : 612.0
Catalogs : 4004.0


Social Networking, Navigation, and Music seem to be the three most attractive based on the table. I would say Social Networking would be the profile we would want to implement.

## 13. Most Popular Apps by Genre on Google Play
We're now going to try to replicate our analysis for the google play data set.

In [63]:
display_table(android_free, 5) # the installs columns

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


Although the numbers are not very precise we don't really need to them to be. We'll assume an app with 1,000,000+ installs has 1,000,000, etc.

In [65]:
# Now we're going repeat what we did with prime_genre but also make
# sure to appropriately convert the strings into int.
categories_android = freq_table(android_free, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in android_free:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

From the table we can deduce that it definitely has to be an entertainment app. We can go further and actually say that it needs to be a social media app. I would recommend a social media app.